File: jvm/core/src/main/java/se/idkollen/client/endpoints/DocumentEndpoint.java
package se.idkollen.client.endpoints;

import se.idkollen.client.PollOptions;
import se.idkollen.client.internal.Transport;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import se.idkollen.client.models.DocumentUploadResponse;

import java.util.concurrent.CompletableFuture;

/** Endpoint for document upload and download operations. */
public class DocumentEndpoint {
    private final Transport transport;

    public DocumentEndpoint(Transport transport) {
        this.transport = transport;
    }

    /**
     * Upload a document for use in signing sessions.
     *
     * @param bytes    Raw file bytes.
     * @param filename File name sent in the multipart form.
     * @param mimeType MIME type of the file.
     * @return Upload response containing the document UUID and hash.
     */
    public CompletableFuture<DocumentUploadResponse> upload(byte[] bytes, String filename, String mimeType) {
        var fileBody = RequestBody.create(bytes, MediaType.get(mimeType));
        var form = new MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", filename, fileBody)
            .build();

        return transport.postMultipart("/v3/document", form, DocumentUploadResponse.class);
    }

    /** Upload a document with the default MIME type {@code application/pdf}. */
    public CompletableFuture<DocumentUploadResponse> upload(byte[] bytes, String filename) {
        return upload(bytes, filename, "application/pdf");
    }

    /**
     * Download a previously uploaded or signed document.
     *
     * @return Raw file bytes.
     */
    public CompletableFuture<byte[]> download(String id) {
        return transport.getBytes("/v3/document/" + id);
    }

    /** Delete a previously uploaded document. */
    public CompletableFuture<Void> delete(String id) {
        return transport.delete("/v3/document/" + id);
    }
}

File: php/src/Endpoints/DocumentEndpoint.php
<?php

declare(strict_types=1);

namespace Idkollen\Client\Endpoints;

use Idkollen\Client\Models\Document\DocumentUploadResponse;
use Idkollen\Client\Transport;

final class DocumentEndpoint
{
    public function __construct(private readonly Transport $transport) {}

    public function upload(string $data, string $filename, string $mimeType = 'application/pdf'): DocumentUploadResponse
    {
        $d = $this->transport->postMultipart('/document', $data, $filename, $mimeType);
        return new DocumentUploadResponse(id: $d['id'], hash: $d['hash']);
    }

    public function download(string $id): string
    {
        return $this->transport->getRaw('/document/' . $id);
    }

    public function delete(string $id): void
    {
        $this->transport->delete('/document/' . $id);
    }
}
